### PR TITLE
[ZF3] [bc-break] remove ControllerLoader,  use ControllerManager instead

### PR DESCRIPTION
**Pull request by samsonasik**
see https://github.com/zendframework/zf2/pull/4962

### CODE REVIEW

**Comment by samsonasik** [comment]
rebased

File: test/Service/ControllerManagerFactoryTest.php [hunk 1]
```diff
@@ -70,25 +70,31 @@ public function testCannotLoadInvalidDispatchable()
...
     {
+<<<<<<< HEAD:test/Service/ControllerLoaderFactoryTest.php
         $this->loader = $this->services->get('ControllerLoader');
         $config = [
             'invokables' => [
+=======
+        $this->manager = $this->services->get('ControllerManager');
+        $config = array(
+            'invokables' => array(
+>>>>>>> remove ControllerLoader, use ControllerManager instead:test/Service/ControllerManagerFactoryTest.php
```
Comment (Maks3w) [inline test/Service/ControllerManagerFactoryTest.php]: Bad rebase

Comment (samsonasik) [inline test/Service/ControllerManagerFactoryTest.php]: fixed.

File: src/Service/ControllerManagerFactory.php [hunk 2]
```diff
@@ -13,7 +13,7 @@
 use Zend\ServiceManager\FactoryInterface;
 use Zend\ServiceManager\ServiceLocatorInterface;
 
-class ControllerLoaderFactory implements FactoryInterface
+class ControllerManagerFactory implements FactoryInterface
```
Comment (Maks3w) [inline src/Service/ControllerManagerFactory.php:16]: Instead create a new class and inherit the old one from the new one. By this way BC is preserved

File: src/Service/ServiceListenerFactory.php [hunk 3]
```diff
@@ -94,8 +94,6 @@ class ServiceListenerFactory implements FactoryInterface
             'Zend\View\Resolver\TemplatePathStack'       => 'ViewTemplatePathStack',
             'Zend\View\Resolver\AggregateResolver'       => 'ViewResolver',
             'Zend\View\Resolver\ResolverInterface'       => 'ViewResolver',
-            'ControllerManager'                          => 'ControllerLoader',
-        ],
```
Comment (Maks3w) [inline src/Service/ServiceListenerFactory.php:98]: Is this correct?

File: src/Service/ControllerManagerFactory.php (hunk 2 above)
Comment (samsonasik) [inline src/Service/ControllerManagerFactory.php:16]: as this is for ZF3, I thik it is ok to remove the old one /cc @weierophinney

File: src/Service/ServiceListenerFactory.php (hunk 3 above)
Comment (samsonasik) [inline src/Service/ServiceListenerFactory.php:98]: yes, at ZF3, the old one ( ControllerLoader ) will no longer used. https://github.com/zendframework/zf2/pull/4962

Comment (samsonasik) [inline src/Service/ServiceListenerFactory.php:98]: I forgot to add back closing "]", fixed now

File: src/Service/ControllerManagerFactory.php (hunk 2 above)
Comment (Maks3w) [inline src/Service/ControllerManagerFactory.php:16]: Do it and this could be merged in 2.6

Comment (samsonasik) [inline src/Service/ControllerManagerFactory.php:16]: done, and <code>ControllerLoaderFactory</code> marked as @deprecated.

File: src/Service/ServiceListenerFactory.php [hunk 4]
```diff
@@ -44,7 +44,7 @@ class ServiceListenerFactory implements FactoryInterface
         'factories' => [
             'Application'                    => 'Zend\Mvc\Service\ApplicationFactory',
             'Config'                         => 'Zend\Mvc\Service\ConfigFactory',
-            'ControllerLoader'               => 'Zend\Mvc\Service\ControllerLoaderFactory',
```
Comment (Maks3w) [inline src/Service/ServiceListenerFactory.php:47]: Old key must be preserved for bc

Comment (Maks3w) [inline src/Service/ServiceListenerFactory.php:47]: Add a comment with `@deprecated` for to remove in the future

File: src/Service/ControllerManagerFactory.php [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+ * @license   http://framework.zend.com/license/new-bsd New BSD License
+ */
+
+namespace Zend\Mvc\Service;
+
+use Zend\Mvc\Controller\ControllerManager;
+use Zend\ServiceManager\FactoryInterface;
+use Zend\ServiceManager\ServiceLocatorInterface;
+
+class ControllerManagerFactory implements FactoryInterface
```
Comment (Maks3w) [inline src/Service/ControllerManagerFactory.php:16]: Add description with the purpose of the class

File: src/Service/ServiceListenerFactory.php (hunk 4 above)
Comment (samsonasik) [inline src/Service/ServiceListenerFactory.php:47]: done

File: src/Service/ControllerManagerFactory.php (hunk 5 above)
Comment (samsonasik) [inline src/Service/ControllerManagerFactory.php:16]: any suggestion description ?

Comment (Maks3w) [inline src/Service/ControllerManagerFactory.php:16]: Following the [SRP](http://en.wikipedia.org/wiki/Single_responsibility_principle) of SOLID describe what is this class responsible

File: src/Service/ControllerLoaderFactory.php [hunk 6]
```diff
@@ -9,40 +9,9 @@
...
-use Zend\ServiceManager\FactoryInterface;
-use Zend\ServiceManager\ServiceLocatorInterface;
-
-class ControllerLoaderFactory implements FactoryInterface
+/**
+ * @deprecated use ControllerManagerFactory instead
+ */
+class ControllerLoaderFactory extends ControllerManagerFactory
 {
-    /**
```
Comment (Maks3w) [inline src/Service/ControllerLoaderFactory.php:18]: Add a constructor with 

`trigger_error('Description', E_USER_DEPRECATED);`

Comment (samsonasik) [inline src/Service/ControllerLoaderFactory.php:18]: done

File: src/Service/ControllerManagerFactory.php (hunk 5 above)
Comment (samsonasik) [inline src/Service/ControllerManagerFactory.php:16]: done

**Comment by samsonasik** [comment]
I've updated PR title

**Comment by samsonasik** [comment]
based on @weierophinney comment on https://github.com/zendframework/zend-authentication/pull/2#issuecomment-110043990 , I will update it as fully remove ControllerLoader, and it will be a PR for 3.0

File: src/Service/ControllerManagerFactory.php (hunk 2 above)
Comment (samsonasik) [inline src/Service/ControllerManagerFactory.php:16]: as https://github.com/zendframework/zend-authentication/pull/2#issuecomment-110043990 , ControllerLoader actually removed for 3.0

File: src/Service/ServiceListenerFactory.php (hunk 4 above)
Comment (samsonasik) [inline src/Service/ServiceListenerFactory.php:47]: now it for 3.0, can be actually removed.

**Comment by samsonasik** [comment]
PR title and commits force updated so we use it for 3.0

**Comment by samsonasik** [comment]
@Ocramius it can be marked as 3.0.0 milestone again ;)

**Comment by Maks3w** [comment]
With the previous approach this could be merged right now. With the v3 tunel vision things will be freeze sine die

**Comment by Maks3w** [comment]
Supersede by #36

**Pull request closed without merging**